src/core/Helpers.cpp

```diff
@@ -143,6 +143,25 @@ bool IsGasWorker::operator()(const sc2::Unit& unit_) const {
     return false;
 }
 
+bool IsTownHall::operator()(const sc2::Unit& unit_) const {
+    return unit_.unit_type == sc2::UNIT_TYPEID::PROTOSS_NEXUS ||
+           unit_.unit_type == sc2::UNIT_TYPEID::TERRAN_COMMANDCENTER ||
+           unit_.unit_type == sc2::UNIT_TYPEID::TERRAN_ORBITALCOMMAND ||
+           unit_.unit_type == sc2::UNIT_TYPEID::TERRAN_PLANETARYFORTRESS ||
+           unit_.unit_type == sc2::UNIT_TYPEID::ZERG_HATCHERY ||
+           unit_.unit_type == sc2::UNIT_TYPEID::ZERG_HIVE ||
+           unit_.unit_type == sc2::UNIT_TYPEID::ZERG_LAIR;
+}
+
+bool IsBuiltTownHall::operator()(const sc2::Unit& unit_) const {
```

Comment:
I would suggest to create a chaining filter class here in order to reuse the code.
i.e.
assuming that we have `IsNonSaturatedTownHall`, `IsTownHall` and `IsFinished` it would be convenient to use e.g. 'AllOf' predicate which uses inside all of these. Ideally, it should be part of the API.

Reply:
I'm having some trouble creating this. Template compiler errors were appearing because std::set tries to create a ::toString() which doesn't work with bool function pointers, and I can't seem to find the right way to code it. Do you see what I am doing wrong?  (code smushed for space)

struct AllOf {
    explicit AllOf(const std::set<sc2::Filter*>& filters_);
    bool operator()(const sc2::Unit& unit_) const;
 private:
    const std::set<sc2::Filter*> m_filters; };

AllOf::AllOf(const std::set<sc2::Filter*>& filters_): m_filters(filters_) {}

bool AllOf::operator()(const sc2::Unit& unit_) const {
    for (auto f : m_filters) {
        if (!(*f)(unit_))
            return false; }
    return true; }

std::set<sc2::Filter*> filters;
filters.insert(sc2::IsTownhall);  //with and without ()'s on IsTownHall
filters.insert(IsNonSaturated);

auto townhalls = gAPI->observer().GetUnits(filters, sc2::Unit::Alliance::Self);

Reply:
The simples way would be to use this approach:
http://www.cplusplus.com/reference/algorithm/all_of/

This lambda should chain our predicates.

Speaking about creating a special class, std::vector is enough, I think.